### PR TITLE
Fusion & Construction file changes (to reflect the fusion changes)

### PR DESCRIPTION
**Pull request by sjjs7**
Will let you know once these have passed holtest!

### CODE REVIEW

File: Constructions/Induction_Schema_Example.ml [hunk 1]
```diff
@@ -1,16 +1,19 @@
+needs "Constructions/epsilon.ml";;
+needs "Constructions/QuotationTactics.ml";;
 (*Testing a few proofs to make sure this definition works*)
 
 (*Trivial proof that x is effecitve in x + 3*)
-prove(effectiveIn `x:num` `x + 3`,
+prove(mk_neg (mk_not_effective_in `x:num` `x + 3` `y:num`),
```
Comment (JacquesCarette) [inline Constructions/Induction_Schema_Example.ml]: Why the change from a positive statement to a double-negative?

File: Constructions/QuotationOperatorTests.ml [hunk 2]
```diff
@@ -1,22 +1,24 @@
-(*Similair to the tests for epsilon, this will prove a few theoremms to verify that _Q_ and Q_ _Q work as intended, along with a few tests with OCaml functions to ensure the Quote term is working correctly*)
+needs "Constructions/QuotationTactics.ml";;
+needs "Constructions/epsilon.ml";;
+(*Similair to the tests for epsilon, this will prove a few theorems to verify that _Q_ and Q_ _Q work as intended, along with a few tests with OCaml functions to ensure the Quote term is working correctly*)
```
Comment (JacquesCarette) [inline Constructions/QuotationOperatorTests.ml]: You might want to fix the typo in the comment while you're at it...

File: fusion.ml [hunk 3]
```diff
@@ -75,6 +75,7 @@ module type Hol_kernel =
       val variant : term list -> term -> term
       val vsubst : (term * term) list -> term -> term
       val qsubst : (term * term) list -> term -> term
+      val BETA_RED_BY_SUB : term -> thm
```
Comment (JacquesCarette) [inline fusion.ml]: This stands out (all caps) amongst things that aren't. Does it belong in this part of the signature?

File: fusion.ml [hunk 4]
```diff
@@ -231,6 +230,8 @@ module Hol : Hol_kernel = struct
           (Tyapp(_,args)) -> itlist (union o tyvars) args []
         | (Tyvar v as tv) -> [tv]
 
+  let is_defined_type ty = if tyvars ty = [] then true else false
```
Comment (JacquesCarette) [inline fusion.ml]: Argh!  Anti-pattern!  This should be
```ocaml
let is_defined_type ty = ityvars ty = []
```

File: fusion.ml [hunk 5]
```diff
@@ -439,6 +442,23 @@ let rec type_subst i ty =
   let dest_eval = 
     function (Eval(e,ty)) -> e,ty | _ -> failwith "dest_eval: not an eval"
 
+(* ------------------------------------------------------------------------- *)
+(* Add to theorem & not-effective lists                                      *)
+(* ------------------------------------------------------------------------- *)
+
+  let addThm tm = match tm with 
+      | Sequent(asl,c) -> if (forall (fun a -> is_proven_thm a) asl) then
```
Comment (JacquesCarette) [inline fusion.ml]: I much prefer to put the nested if on the next line, and indent it 4 from the `|` (which is 2 from the `Sequent`). That way the code isn't so wide, with lots of wasted space on the left.

File: fusion.ml [hunk 6]
```diff
@@ -673,29 +681,36 @@ let rec type_subst i ty =
...
+                    (match ilist with
+                      | [(t,x)] -> Comb(Abs(x,tm),t)
+                      | _ -> failwith "More than one substitution into an abstraction with a resolved substitution.") in 
     fun theta ->
       if theta = [] then (fun tm -> tm) else
       if forall (function (t,Var(_,y)) -> Pervasives.compare (type_of t) y = 0
                         | _ -> false) theta
       then vsubst theta else failwith "vsubst: Bad substitution list"
 
+  let BETA_RED_BY_SUB tm = if is_comb tm && is_abs(fst(dest_comb tm))
```
Comment (JacquesCarette) [inline fusion.ml]: Same here: the `if` should be on the next line, and indented just 2.

File: fusion.ml [hunk 7]
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
...
       | other when type_of(other) = Tyapp("epsilon",[]) -> Hole(other,type_of other)
       | _ -> failwith "constructionToTerm"
 
-  let TERM_TO_CONSTRUCTION tm = match tm with
-      |  Quote(exp,t) when type_of exp = t -> Sequent([],safe_mk_eq tm (termToConstruction exp))
-      |  Quote(_,_) -> failwith "TERM_TO_CONSTRUCTION: BAD QUOTE"
-      | _ -> failwith "TERM_TO_CONSTRUCTION"
-  
-  let CONSTRUCTION_TO_TERM tm = try Sequent([],safe_mk_eq tm (mk_quote (constructionToTerm tm))) with Failure _ -> failwith "CONSTRUCTION_TO_TERM"
+  let is_proper_construc c = if can constructionToTerm c then true else false
```
Comment (JacquesCarette) [inline fusion.ml]: anti-pattern

File: fusion.ml [hunk 8]
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
...
-  
-  let CONSTRUCTION_TO_TERM tm = try Sequent([],safe_mk_eq tm (mk_quote (constructionToTerm tm))) with Failure _ -> failwith "CONSTRUCTION_TO_TERM"
+  let is_proper_construc c = if can constructionToTerm c then true else false 
+
+  let proper_e_term tm = is_proper_construc tm || is_quote tm
+
+  let LAW_OF_QUO tm = match tm with
+    | Var(_,_) | Const(_,_) | Comb(_,_) | Abs(_,_) | Quote(_,_) | Hole(_,_) -> 
+      if is_eval_free tm then Sequent([],safe_mk_eq (mk_quote tm) (termToConstruction tm))
+      else failwith "Not eval free"
```
Comment (JacquesCarette) [inline fusion.ml]: This is a bad error message. You should be able to print the occurrence of `Eval` that causes the problem. This means that `is_eval_free` shouldn't produce a boolean, but rather a data-type with 2 constructors (call them `Ok` and `Problem` where `Problem` contains the payload necessary for you to print a decent error.

File: fusion.ml [hunk 9]
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
...
+    | Eval(_,_) -> failwith "Not eval free"
```
Comment (JacquesCarette) [inline fusion.ml]: Here, giving a better error is easy.

File: fusion.ml [hunk 10]
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
...
+    | _ -> failwith "Not a proper term"
```
Comment (JacquesCarette) [inline fusion.ml]: anti-pattern: catch-all match.

File: fusion.ml [hunk 11]
```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
...
   the relevant function to it*)
 
-  let rec TERM_TO_CONSTRUCTION_CONV tm = match tm with
-    | Const(a,b) -> failwith "TERM_TO_CONSTRUCTION_CONV"
-    | Quote(_,_) -> TERM_TO_CONSTRUCTION tm
-    | Comb(a,b) -> try TERM_TO_CONSTRUCTION_CONV a with Failure _ -> try TERM_TO_CONSTRUCTION_CONV b with Failure _ -> failwith "TERM_TO_CONSTRUCTION_CONV"
-    | _ -> failwith "TERM_TO_CONSTRUCTION_CONV"
+  let LAW_OF_QUO_CONV tm = LAW_OF_QUO tm
+
+  let rec QUOTE_TO_CONSTRUCTION_CONV tm = match tm with
```
Comment (JacquesCarette) [inline fusion.ml]: you can use `function` for this pattern, i.e.
```ocaml
let foo xx = match xx with
```
is just
```ocaml
let foo = function
```
(there are more places where this can be used)

File: fusion.ml [hunk 12]
```diff
@@ -1212,75 +1244,73 @@ let rec type_subst i ty =
...
-
-  let rec EVAL_QUOTE_CONV tm = match tm with
-    | Comb(a,b) -> (try (EVAL_QUOTE_CONV a) with Failure _ -> try (EVAL_QUOTE_CONV b) with Failure _ -> failwith "EVAL_QUOTE_CONV")
-    | Abs(a,b) -> (try (EVAL_QUOTE_CONV a) with Failure _ -> try (EVAL_QUOTE_CONV b) with Failure _ -> failwith "EVAL_QUOTE_CONV")
-    | Eval(e,ty) -> EVAL_QUOTE tm
-    | _ -> failwith "EVAL_QUOTE_CONV"
+      | _ -> failwith "LAW_OF_DISQUO: Not evaluating a quotation!"
 
+  let rec LAW_OF_DISQUO_CONV tm = match tm with
+    | Comb(a,b) -> (try (LAW_OF_DISQUO_CONV a) with Failure _ -> try (LAW_OF_DISQUO_CONV b) with Failure _ -> failwith "LAW_OF_DISQUO_CONV")
```
Comment (JacquesCarette) [inline fusion.ml]: Same comments about error messages.

File: fusion.ml [hunk 13]
```diff
@@ -1212,75 +1244,73 @@ let rec type_subst i ty =
...
-    let conclud = safe_mk_eq (Eval(Comb(Comb(Const("app",makeHolFunction (makeHolType "epsilon" []) (makeHolFunction (makeHolType "epsilon" []) (makeHolType "epsilon" []))),tm1),tm2),((type_of tm1)))) (Comb(Eval(tm1,makeHolFunction (type_of tm1) (type_of tm2)),Eval(tm2,(type_of tm2)))) in
-                           Sequent([], internal_make_imp anticed conclud)
+  let is_expr_ty = Const("isExprType",makeHolFunction (makeHolType "epsilon" []) (makeHolFunction (makeHolType "type" []) (makeHolType "bool" [])))
+
+  let QUO_DISQUO tm = if proper_e_term tm then  
+                        let iet = Comb(Comb(is_expr_ty,tm),matchType (Tyapp("epsilon",[]))) in
+                        Sequent([],(internal_make_imp iet (safe_mk_eq (Eval(Comb(Const("Quo",makeHolFunction (makeHolType "epsilon" []) (makeHolType "epsilon" [])),tm),Tyapp("epsilon",[]))) tm )))
+                      else failwith "QUO_DISQUO"
+
+  let ABS_DISQUO var tm = if not (is_var var) then failwith "ABS_DISQUO: not a variable"
```
Comment (JacquesCarette) [inline fusion.ml]: I have a feeling that this could be written more clearly with pattern matching than via a whole bunch of if-then.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: Constructions/Induction_Schema_Example.ml (hunk 1 above)
Comment (sjjs7) [reply Constructions/Induction_Schema_Example.ml]: Yes I guess that does look off- it's because I changed the default from effective-in to not-effective-in. I should still have an effective-in statement though, will add.

File: fusion.ml (hunk 4 above)
Comment (sjjs7) [reply fusion.ml]: Yes that's awful!

**Comment by sjjs7** [comment]
Ok I think I fixed the issues, let me know what you think. 
(I still have to run holtest)

File: fusion.ml [hunk 14]
```diff
@@ -313,49 +327,48 @@ let rec type_subst i ty =
...
   (*Checks if a term is eval-free*)
-  let rec is_eval_free tm = match tm with
-    | Var(_,_) -> true
-    | Const(_,_) -> true
-    | Comb(a,b) -> is_eval_free a && is_eval_free b
-    | Abs(a,b) -> is_eval_free a && is_eval_free b
+  let rec is_eval_free tm = match tm with 
+    | Var(_,_) -> Ok
+    | Const(_,_) -> Ok
+    | Comb(a,b) -> if is_eval_free a = Ok then is_eval_free b else is_eval_free a
```
Comment (JacquesCarette) [inline fusion.ml]: You want to let-bind `is_eval_free a` so that you can re-use it in the `else` clause.  [If Ocaml had monads, it would be even simpler, but...]

File: fusion.ml [hunk 15]
```diff
@@ -417,11 +431,14 @@ let rec type_subst i ty =
         -> Comb(f,a)
     | _ -> failwith "mk_comb: types do not agree"
 
-  let mk_quote t = if is_eval_free t then Quote(t,qcheck_type_of t) else failwith "Can only quote eval-free terms"
+  let mk_quote t = 
+     match (is_eval_free t) with 
+        | Ok -> Quote(t,qcheck_type_of t)
+        | Issue tm -> raise (Problem("Cannot quote terms with evaluations. This term contains one:", tm))
```
Comment (JacquesCarette) [inline fusion.ml:436]: Definitely better.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

**Review by JacquesCarette — APPROVED** [review]

**Comment by sjjs7** [comment]
I guess I better check these changes as well...will let you know.

**Comment by sjjs7** [comment]
Alright, it's fine, we're good to go.

**Review by JacquesCarette — APPROVED** [review]
